### PR TITLE
Revert "Redefine scheduler to set learning rate using recursive formula" #14010

### LINKED ISSUE #20527: StepLR, MultiStepLR, ExponentialLR and CosineAnnealingLR scheduler wrong lr value
When the StepLR, MultiStepLR, ExponentialLR or CosineAnnealingLR scheduler is called with the same epoch parameter the optimizer value is further reduced even though it's the same epoch

a sample code
```
import torch.optim as optim
from torch import nn

conv = nn.Conv2d(3,3,3)
optimizer = optim.Adam(conv.parameters()) 
lr_scheduler = optim.lr_scheduler.StepLR(optimizer, 2)
```
now if we call `lr_scheduler.step(epoch=2)` multiple times 
```
for _ in range(10):
    lr_scheduler.step(2)
    print(optimizer.param_groups[0]['lr'])
```
output:
```
>>> 0.0001
>>> 1e-05
>>> 1.0000000000000002e-06
>>> 1.0000000000000002e-07
>>> 1.0000000000000004e-08
>>> 1.0000000000000005e-09
>>> 1.0000000000000006e-10
>>> 1.0000000000000006e-11
>>> 1.0000000000000006e-12
>>> 1.0000000000000007e-13
```

Even if such use-case is bizarre, this is extremely unexpected.
This is happens using PyTorch version 1.1.0 but not 1.0.1. Because pull(#14010) redefined StepLR, MultiStepLR, ExponentialLR and CosineAnnealingLR to directly use the learning rate variable of the optimizer rather than using `base_lr` defined by `_LRScheduler`. This was in order to support multiple simultaneous schedulers (#13022)!

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#21463 Revert "Redefine scheduler to set learning rate using recursive formula" #14010**

This revert was done by hand, by replacing the recursive get_lr definitions with
the non-recursive ones and then deleting tests that failed.  I did NOT delete
the "trivial" legacy implementation tests; they might still be useful if we
attempt this patch again.

Fixes #20527

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D15747426](https://our.internmc.facebook.com/intern/diff/D15747426)

### CODE REVIEW

**Comment by ezyang** [comment]
cc @jasam-sheja @jafermarq

@soumith We should consider including this revert in a point release, if we decide to do it.

**Comment by ezyang** [comment]
@fmassa  says this will break torchvision code if we revert it.

**Comment by fmassa** [comment]
I can update torchvision to use my own lr_schedulers to bypass a potential revert, like what I did in [maskrcnn-benchmark in the past](https://github.com/facebookresearch/maskrcnn-benchmark/blob/c8eff2ca2d5c1c14393f1a1b21853895eaacfd51/maskrcnn_benchmark/solver/lr_scheduler.py#L7-L9), but it would be great to have a way of composing lr schedulers natively supported in PyTorch.

**Review by zou3519 — COMMENTED** [review]
there are changes to third_party/fbgemm

File: test/test_optim.py [hunk 1]
```diff
@@ -744,141 +744,6 @@ def test_reduce_lr_on_plateau8(self):
                                       threshold=0.1, patience=5, cooldown=5)
         self._test_reduce_lr_on_plateau(scheduler, targets, metrics, epochs)
 
-    def test_compound_step_and_multistep_lr(self):
```
Comment (zou3519) [inline test/test_optim.py]: I'm surprised that these are the only tests that failed; #14010 added a lot of things into test_optim.py

**Review by zou3519 — APPROVED** [review]

File: test/test_optim.py (hunk 1 above)
Comment (ezyang) [reply test/test_optim.py]: Yeah, the other tests were just comparing against reference implementations (which still works; it's just a less interesting test.)
